kotlin/iwa-sample-app/src/main/java/com/esri/android/samples/iwasampleapp/MainActivity.kt

```diff
@@ -0,0 +1,173 @@
+                } else {
+                    // Portal is failed to load. Handler load errors
+                    val error = portal.loadError
+                    if (error.errorCode == 17) {
+                        // User canceled exception
+                        Toast.makeText(applicationContext, "Portal loading is cancelled by user!", Toast.LENGTH_SHORT).show()
+                    } else {
+                        // Other failures
+                        Toast.makeText(applicationContext, 
+                                "Portal fails to load: " + error.cause?.message ?: error.message, Toast.LENGTH_SHORT).show()
+                    }
+                }
+            }
+        }
+    }
+}
+
+const val MAX_ATTEMPTS = 5
+
+/**
+ * A custom authentication challenge handler to handle user credential challenge.
+ * 
+ * When a user credential challenge is issues a dialog will be presented for credential.
+ * The portal url will be displayed as message in the dialog. If wrong credential has been
+ * passed in the previous attempt, a different message will be display in the dialog.
+ * The dialog has two edit text boxes for username and password respectively. Other SDKs'
+ * sample may have one more parameter for IWA domain. As indicated by the Javadoc of UseCredential
+ * our SDK is in favor of passing username as username@domain. This sample doesn't provide
+ * a edit text box for domain to simplify the logic.
+ *  
+ * @see <a href="https://developers.arcgis.com/android/latest/api-reference/reference/com/esri/arcgisruntime/security/UserCredential.html#UserCredential(java.lang.String,%20java.lang.String)">Javadoc of UserCredential</a>
+ * 
+ */
+class IWACustomChallengeHandler(val activity: Activity) : AuthenticationChallengeHandler {
+    override fun handleChallenge(challenge: AuthenticationChallenge): AuthenticationChallengeResponse {
+        if (challenge.type == AuthenticationChallenge.Type.USER_CREDENTIAL_CHALLENGE) {
+            if (challenge.failureCount > MAX_ATTEMPTS) {
+                // Exceed maximum amount of attempts. Act like it was a cancel
+                Toast.makeText(activity, "Exceed maximum amount of attempts. Please try again!", Toast.LENGTH_SHORT).show()
+                return AuthenticationChallengeResponse(AuthenticationChallengeResponse.Action.CANCEL, challenge)
+            }
+            
+            // Create a countdown latch with a count of one to synchronize the dialog
+            val signal : CountDownLatch = CountDownLatch(1)
+            var credential : UserCredential?
+            credential = null
+            
+            // Present the sign-in dialog
+            activity.runOnUiThread(java.lang.Runnable { 
+                // Inflate the layout
+                val dialogView = activity.layoutInflater.inflate(R.layout.credential_dialog, null)
+                // Create the dialog
+                val builder: AlertDialog.Builder = AlertDialog.Builder(activity)
```

Comment:
Should we still be using AlertDialog? What do we se in other samples?

Reply:
other samples are using AlertDialog